### ISSUE TITLE
Assigning value to container-typed field does not validate if the value is a Node

### ISSUE THREAD

**Issue by Jasha10** [omry/omegaconf#747]
**Describe the bug**
If a field in a `BaseContainer` has type `List[...]` or `Dict[..., ...]`, assignment to that field is not validated if the value is an instance of `Node`.

**To Reproduce**
Consider the following example:
```python
from dataclasses import dataclass
from omegaconf import *
from typing import *

@dataclass
class SubscriptedDict:
    dict: Dict[str, int]

cfg = OmegaConf.structured(SubscriptedDict)
# cfg.dict = {"a": "xyz"}  # raises ValidationError, as expected
cfg.dict = DictConfig({"a": "xyz"})  # does not raise!
```
This applies to lists too:
```python
@dataclass
class SubscriptedList:
    list: List[int]
cfg = OmegaConf.structured(SubscriptedList([1,2]))
# cfg.list = ["xyz"]  # raises ValidationError, as expected
cfg.list = ListConfig(["xyz"])  # does not raise an error!
```

**Expected behavior**
I expect the same validation to occur whether the assigned value is a node or not.

**Additional context**
- OmegaConf version: master branch
- Python version: 3.9

**Comment by omry** [comment]
The problem is that we are not re-constructing the input if it's already a DictConfig or a ListConfig.
Solving it, especially efficiently - could be difficult.
This is not a high priority issue right now.

**Comment by Jasha10** [comment]
This bug is fixed by #890.
